src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java

```diff
@@ -224,4 +284,67 @@ public String[] getAllQueryIds() {
+    * The solutions will then be instantiated into the query pattern.
+    * The result may look like the following:<br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "123"}</code><br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "12"}</code><br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/2&gt; ?o . ?o &lt;http://exa.com&gt; "1234"}</code><br/>
+    */
+    private static List<String> instantiatePatternQueries(QuerySource querySource, Config.Pattern config) throws IOException {
+        final var patternQueries = new FileCachingQueryList(querySource);
+        final Pattern pattern = Pattern.compile("%%var\\d+%%");
+        final var instances = new ArrayList<String>();
+        for (int i = 0; i < patternQueries.size(); i++) {
+            // replace all variables in the query pattern with SPARQL variables
+            // and store the variable names
+            var patternQueryString = patternQueries.getQuery(i);
+            final Matcher matcher = pattern.matcher(patternQueryString);
+            final var variables = new ArrayList<String>();
+            while (matcher.find()) {
+                final var match = matcher.group();
+                final var variable = "?" + match.replaceAll("%%", "");
+                variables.add(variable);
+                patternQueryString = patternQueryString.replaceAll(match, variable);
+            }
+
+            // build SELECT query for finding bindings for the variables
+            final var patternQuery = QueryFactory.create(patternQueryString);
+            final var whereClause = "WHERE " + patternQuery.getQueryPattern();
+            final var selectQueryString = new ParameterizedSparqlString();
+            selectQueryString.setCommandText("SELECT DISTINCT " + String.join(" ", variables));
+            selectQueryString.append(" " + whereClause);
+            selectQueryString.append(" LIMIT " + config.limit());
+            selectQueryString.setNsPrefixes(patternQuery.getPrefixMapping());
+
+            // send request to SPARQL endpoint and instantiate the pattern based on results
+            try (QueryExecution exec = QueryExecutionFactory.createServiceRequest(config.endpoint().toString(), selectQueryString.asQuery())) {
+                ResultSet resultSet = exec.execSelect();
+                if (!resultSet.hasNext()) {
+                    LOGGER.warn("No results for query pattern: {}", patternQueryString);
+                }
+                while (resultSet.hasNext()) {
+                    var instance = new ParameterizedSparqlString(patternQueryString);
+                    QuerySolution solution = resultSet.next();
+                    for (String var : resultSet.getResultVars()) {
+                        instance.clearParam(var);
+                        instance.setParam(var, solution.get(var));
+                    }
+                    instances.add(instance.toString());
```

Comment:
We don't need more than limit instances. So, you can break the while look when that many were found.